cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/batch/BatchSourceContext.java

```diff
@@ -42,5 +42,5 @@
   /**
    * @return maximum number of records to read in preview mode.
    */
-  int getMaxPreviewRecords();
+  int getMaxPreviewRecords(int maxPreviewRecordsLimit);
```

Comment:
Didn't realize this method is part of API. Let's not change this else it will be backward incompatible. Instead what we can do is in `MapReduceBatchContext` constructor, make a copy of stageSpec with number of preview records equal to min of what we get from existing stage spec and from data tracer and pass that copied stageSpeck to the base class constructors.